# OpenAI API Core Features Walkthrough

This notebook provides a comprehensive tour of the OpenAI API's key capabilities including text generation, structured outputs, vision, audio, function calling, reasoning models, and embeddings.

In [1]:
# OpenAI API Examples Notebook
# 
# This notebook demonstrates key features of the OpenAI API for developers.
# It provides practical examples of the most important capabilities with executable code.

# First, let's set up our environment by installing the OpenAI Python library
# DEPENDENCY: pip install openai numpy matplotlib pydantic --upgrade --quiet
# (packages should be pre-installed in venv before running this script)

In [ ]:
import os
import json
from getpass import getpass
import numpy as np
import requests
from openai import OpenAI
from pydantic import BaseModel

# API Key Setup - uses environment variable or prompts for input
if not os.environ.get("OPENAI_API_KEY"):
    os.environ["OPENAI_API_KEY"] = getpass("Enter your OpenAI API key: ")

# Client and Model Configuration
client = OpenAI()
MODEL = "gpt-5.2"

In [4]:
import os

# Initialize the client
# Note: In a real application, you would use an environment variable or secure method
# to store your API key. This is just for demonstration.
client = OpenAI(
    # Replace with your actual API key or use: api_key=os.environ.get("OPENAI_API_KEY")
    api_key=os.environ.get("OPENAI_API_KEY")
)

In [5]:
prompt = "Write a one-sentence bedtime story about a unicorn."

response = client.responses.create(
    model=MODEL,
    input=prompt
)

print(f"Prompt: {prompt}")
print(f"Response: {response.output_text}")

# Show how to control generation with parameters
print("\nWith controlled parameters:")
response = client.responses.create(
    model=MODEL,
    input=prompt,
    temperature=0.7,  # Lower for more deterministic outputs
    top_p=0.9         # Nucleus sampling
)

print(f"Response (controlled): {response.output_text}")

Prompt: Write a one-sentence bedtime story about a unicorn.
Response: As the moonlight sparkled on her shimmering mane, Luna the unicorn whispered her dreams to the stars, who eagerly twinkled back in encouragement as she drifted off to sleep.

With controlled parameters:


Response (controlled): As the moonlight danced on the shimmering lake, a gentle unicorn named Luna spread her magical wings and soared into the starry sky, spreading dreams of wonder to all the children asleep below.


---

## 2. Structured Outputs

In [6]:
import json

response = client.responses.create(
   model=MODEL,
    input=[
        {"role": "system", "content": "You are a UI generator AI. Convert the user input into a UI."},
        {"role": "user", "content": "Make a User Profile Form"}
    ],
    text={
        "format": {
            "type": "json_schema",
            "name": "ui",
            "description": "Dynamically generated UI",
            "schema": {
                "type": "object",
                "properties": {
                    "type": {
                        "type": "string",
                        "description": "The type of the UI component",
                        "enum": ["div", "button", "header", "section", "field", "form"]
                    },
                    "label": {
                        "type": "string",
                        "description": "The label of the UI component, used for buttons or form fields"
                    },
                    "children": {
                        "type": "array",
                        "description": "Nested UI components",
                        "items": {"$ref": "#"}
                    },
                    "attributes": {
                        "type": "array",
                        "description": "Arbitrary attributes for the UI component, suitable for any element",
                        "items": {
                            "type": "object",
                            "properties": {
                              "name": {
                                  "type": "string",
                                  "description": "The name of the attribute, for example onClick or className"
                              },
                              "value": {
                                  "type": "string",
                                  "description": "The value of the attribute"
                              }
                          },
                          "required": ["name", "value"],
                          "additionalProperties": False
                      }
                    }
                },
                "required": ["type", "label", "children", "attributes"],
                "additionalProperties": False
            },
            "strict": True,
        },
    },
)

ui = json.loads(response.output_text)


In [7]:
ui

{'type': 'form',
 'label': 'User Profile Form',
 'children': [{'type': 'header',
   'label': 'User Profile Details',
   'children': [],
   'attributes': []},
  {'type': 'field',
   'label': 'First Name',
   'children': [],
   'attributes': [{'name': 'type', 'value': 'text'},
    {'name': 'placeholder', 'value': 'Enter your first name'},
    {'name': 'required', 'value': ''}]}],
 'attributes': [{'name': 'id', 'value': 'user-profile-form'}]}

---

## 2.1 Structured Outputs with Pydantic models

In [8]:
from pydantic import BaseModel

class Step(BaseModel):
    explanation: str
    output: str

class MathReasoning(BaseModel):
    steps: list[Step]
    final_answer: str

# Get the schema and add additionalProperties: false for strict mode
math_schema = MathReasoning.model_json_schema()
math_schema["additionalProperties"] = False
# Also fix nested $defs
if "$defs" in math_schema:
    for def_name, def_schema in math_schema["$defs"].items():
        if "additionalProperties" not in def_schema:
            def_schema["additionalProperties"] = False

completion = client.responses.create(
    model=MODEL,
    input=[
        {"role": "system", "content": "You are a helpful math tutor. Guide the user through the solution step by step."},
        {"role": "user", "content": "how can I solve 8x + 7 = -23"}
    ],
    text={
        "format": {
            "type": "json_schema",
            "name": "math_reasoning",
            "description": "Step-by-step math reasoning with final answer",
            "schema": math_schema,
            "strict": True,
        },
    },
)

# Parse the JSON response into the Pydantic model
math_reasoning = MathReasoning.model_validate_json(completion.output_text)
print(math_reasoning)

steps=[Step(explanation='Start with the original equation: 8x + 7 = -23.', output='8x + 7 = -23'), Step(explanation='Subtract 7 from both sides to isolate the term with x: 8x + 7 - 7 = -23 - 7.', output='8x = -30'), Step(explanation='Now, divide both sides by 8 to solve for x: x = -30 / 8.', output='x = -3.75 or x = -15/4.')] final_answer='x = -3.75'


In [9]:
math_reasoning.steps[0]

Step(explanation='Start with the original equation: 8x + 7 = -23.', output='8x + 7 = -23')

---

## 3. Multimodal Capabilities - Vision

In [10]:
# For notebook demonstration, we'll use a placeholder URL
image_url = "https://images.unsplash.com/photo-1579546929518-9e396f3cc809?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxleHBsb3JlLWZlZWR8MXx8fGVufDB8fHx8&w=1000&q=80"

response = client.responses.create(
    model=MODEL,
    input=[{
        "role": "user",
        "content": [
            {"type": "input_text", "text": "what's in this image?"},
            {
                "type": "input_image",
                "image_url": image_url,
            },
        ],
    }],
)

print(response.output_text)


The image appears to be an abstract gradient with a blend of soft colors, including shades of pink, orange, and blue. It creates a smooth, blended effect, evoking a calming and vibrant aesthetic.


---

## 4. Audio Capabilities - Text to Speech

In [11]:
speech_file_path = "speech.mp3"

with client.audio.speech.with_streaming_response.create(
    model="gpt-5.2-tts",
    voice="coral",
    input="Today is a wonderful day to build something people love!",
    instructions="Speak in a cheerful and positive tone.",
) as response:
    response.stream_to_file(speech_file_path)

---

## 4.1 Audio Capabilities - Speech to Text

In [12]:
audio_file= open("speech.mp3", "rb")

transcription = client.audio.transcriptions.create(
    model="whisper-1", 
    file=audio_file
)

print(transcription.text)

Today is a wonderful day to build something people love.


---

## 5. Function Calling

In [13]:
import requests

def get_weather(latitude, longitude):
    response = requests.get(f"https://api.open-meteo.com/v1/forecast?latitude={latitude}&longitude={longitude}&current=temperature_2m,wind_speed_10m&hourly=temperature_2m,relative_humidity_2m,wind_speed_10m")
    data = response.json()
    return data['current']['temperature_2m']

tools = [{
    "type": "function",
    "name": "get_weather",
    "description": "Get current temperature for provided coordinates in celsius.",
    "parameters": {
        "type": "object",
        "properties": {
            "latitude": {"type": "number"},
            "longitude": {"type": "number"}
        },
        "required": ["latitude", "longitude"],
        "additionalProperties": False
    },
    "strict": True
}]

input_messages = [{"role": "user", "content": "What's the weather like in Paris today?"}]

response = client.responses.create(
    model="gpt-5.2",
    input=input_messages,
    tools=tools,
)

In [14]:
# Extract the tool call and arguments
tool_call = response.output[0]
args = json.loads(tool_call.arguments)
# Call the function
result = get_weather(args["latitude"], args["longitude"])

In [15]:
# Append the tool call and result to the input messages
input_messages.append(tool_call) # append model's function call message
input_messages.append({ # append result message
    "type": "function_call_output",
    "call_id": tool_call.call_id,
    "output": str(result)
})

response_2 = client.responses.create(
    model="gpt-5.2",
    input=input_messages,
    tools=tools,
)
print(response_2.output_text)

The current temperature in Paris is 5°C. If you need more details or a forecast, let me know!


---

## 6. Reasoning Models

In [16]:
# response_o1 = client.responses.create(
#     model="o1",
#     input="Design an algorithm to find the shortest path in a graph."
# )

# print("o1 Response:")
# print(response_o1.output_text)

# Using o4-mini for faster reasoning
response_o4 = client.responses.create(
    model="o4-mini",
    input="Explain how to implement a hash table."
)

print("\no3-mini Response:")
print(response_o4.output_text)


o3-mini Response:
A hash table is a data structure that provides (amortized) O(1) insert, lookup, and delete by mapping keys to array indices via a hash function and handling collisions. Here’s how you can implement one:

1. Choose your collision‐resolution strategy  
   a. Separate chaining (each bucket holds a linked list or dynamic array of entries)  
   b. Open addressing (all entries live in one array; collisions are resolved by probing)

2. Define your basic data structures  
   For separate chaining:  
     struct Entry {  
       KeyType key;  
       ValueType value;  
       Entry* next;  
     };  
     // The table is an array of Entry* (bucket heads)  
     Entry* buckets[N];  

   For open addressing:  
     enum SlotState { EMPTY, OCCUPIED, DELETED };  
     struct Slot {  
       KeyType key;  
       ValueType value;  
       SlotState state;  
     };  
     // The table is an array of Slot  
     Slot table[N];  

3. Hash function  
   // Must produce an integer in 

In [17]:
## With reasoning models, you can view the reasoning process and the steps taken to arrive at the answer.
print("This is the reasoning configuration for o4-mini:", response_o4.reasoning)

This is the reasoning configuration for o4-mini: Reasoning(effort='medium', generate_summary=None, summary=None)


## 7. Embeddings

## Summary

- **Text Generation**: Use `client.responses.create()` with `input` parameter and control output with `temperature` and `top_p`
- **Structured Outputs**: Define JSON schemas or use Pydantic models with `text.format` to get typed, validated responses
- **Multimodal**: Vision (image analysis), text-to-speech, and speech-to-text capabilities are all available through the API
- **Function Calling**: Define tools with JSON schemas and handle tool calls in an agentic loop
- **Embeddings**: Generate vector representations of text for semantic search and similarity comparisons

In [18]:
import numpy as np

# First, let's create embeddings for a set of words
words = [
    "king", "queen", "man", "woman", 
    "apple", "banana", "orange", "pear",
    "castle", "throne"
]

# Get embeddings for all words
response = client.embeddings.create(
    model="text-embedding-3-large",
    input=words,
    encoding_format="float"
)

# Extract the embeddings
embeddings = [data.embedding for data in response.data]

print(f"Embedding dimension: {len(embeddings[0])}")
print(f"Number of embeddings: {len(embeddings)}")

# Function to compute dot product between two vectors
def dot_product(vec1, vec2):
    return np.dot(vec1, vec2)

# Compute similarity matrix (dot products between all pairs)
similarity_matrix = np.zeros((len(words), len(words)))
for i in range(len(words)):
    for j in range(len(words)):
        similarity_matrix[i][j] = dot_product(embeddings[i], embeddings[j])

# Print similarity matrix with labels
print("\nSimilarity Matrix (Dot Products):")
print("          " + " ".join(f"{word:<8}" for word in words))
for i, word in enumerate(words):
    row_values = " ".join(f"{similarity_matrix[i][j]:.4f}  " for j in range(len(words)))
    print(f"{word:<10} {row_values}")

# Check specific relationships
king_queen_similarity = dot_product(embeddings[0], embeddings[1])
apple_banana_similarity = dot_product(embeddings[4], embeddings[5])

print("\nSpecific relationships:")
print(f"Similarity between 'king' and 'queen': {king_queen_similarity:.4f}")
print(f"Similarity between 'apple' and 'banana': {apple_banana_similarity:.4f}")
print(f"Similarity between 'king' and 'apple': {dot_product(embeddings[0], embeddings[4]):.4f}")

# We expect king/queen to be closer to each other than king/apple
# And apple/banana to be closer to each other than queen/banana

Embedding dimension: 3072
Number of embeddings: 10

Similarity Matrix (Dot Products):
          king     queen    man      woman    apple    banana   orange   pear     castle   throne  
king       1.0000   0.5523   0.4121   0.2923   0.3221   0.3272   0.2869   0.2760   0.3564   0.3967  
queen      0.5523   1.0000   0.3072   0.4131   0.3145   0.3191   0.2983   0.2995   0.2970   0.3353  
man        0.4121   0.3072   1.0000   0.5714   0.3098   0.3497   0.2973   0.2695   0.2999   0.2649  
woman      0.2923   0.4131   0.5714   1.0000   0.3199   0.2938   0.2783   0.2533   0.2450   0.2490  
apple      0.3221   0.3145   0.3098   0.3199   1.0000   0.4620   0.4588   0.4390   0.3003   0.2340  
banana     0.3272   0.3191   0.3497   0.2938   0.4620   1.0000   0.4580   0.3636   0.2779   0.2075  
orange     0.2869   0.2983   0.2973   0.2783   0.4588   0.4580   1.0000   0.3822   0.2849   0.2173  
pear       0.2760   0.2995   0.2695   0.2533   0.4390   0.3636   0.3822   1.0000   0.2789   0.1929  
castle

### Interpreting the Embedding Similarities

**Key observations from the similarity matrix:**

| Comparison | Similarity | Interpretation |
|------------|------------|----------------|
| king - queen | **0.55** | Strong relationship (royalty, leadership) |
| man - woman | **0.57** | Strong relationship (human, gender pair) |
| apple - banana | **0.46** | Moderate relationship (both fruits) |
| king - apple | **0.32** | Weak relationship (unrelated concepts) |
| castle - throne | **0.39** | Moderate relationship (royalty context) |

**What this tells us:**
1. **Semantic relationships are captured**: Word pairs with real-world relationships (king/queen, man/woman) have higher similarity scores
2. **Category grouping works**: Fruits cluster together (apple-banana: 0.46, apple-orange: 0.46)
3. **Cross-category similarity is low**: Unrelated concepts like "king" and "apple" show low similarity (0.32)

**Practical application:** These semantic relationships enable:
- **Semantic search**: Find documents about "monarch" even when searching for "king"
- **Clustering**: Group similar items without explicit labels
- **Analogies**: The famous "king - man + woman = queen" works because these relationships are encoded in the vector space